File: projeto-haskell/src/Controller/QuestionController.hs
{-# LANGUAGE OverloadedStrings #-}
module Controller.QuestionController where
import Entities.Question
import Database.SQLite.Simple
import Database.SQLite.Simple.FromRow
import Database.SQLite.Simple.ToField
import Utils.Util
import Database.SQLite.Simple.Types
import Entities.Answer
import Data.List (intercalate, intersperse)

instance FromRow Question where
  fromRow = Question <$> field
                     <*> field
                     <*> field
                     <*> field
                     <*> field
                     <*> field
                     <*> field
instance FromRow Answer where
  fromRow = Answer <$> field
                   <*> field
                   <*> field

addQuestion :: String -> Int -> Int -> Int -> String -> IO String
addQuestion formulation difficulty duration typeQuestion quizId = do
    conn <- open dbPath
    uuidQuestion <- getRandomUUID
    execute conn "INSERT INTO question (question_id,formulation,difficulty,\
        \time,type_question,quiz_id) VALUES (?,?,?,?,?,?)"
        [uuidQuestion,formulation,show difficulty,show duration,show typeQuestion,quizId]
    return uuidQuestion

getAllQuestions :: String -> IO [Question]
getAllQuestions quiz_id = do
    conn <- open dbPath
    query conn "SELECT * from question WHERE quiz_id = ?"  (Only quiz_id)
        :: IO [Question]

getAllAnswers :: String -> IO [Answer]
getAllAnswers question_id = do
    conn <- open dbPath
    query conn "SELECT * from answer WHERE question_id = ?"  (Only question_id)
        :: IO [Answer]

updateQuestionRightAnswer:: String -> String -> IO()
updateQuestionRightAnswer question_id right_answer = do
    conn <- open dbPath
    execute conn "UPDATE question SET right_answer = ? WHERE question_id = ?"
        (right_answer, question_id)

updateQuestion:: Question -> IO()
updateQuestion question = do
      conn <- open dbPath
      execute conn "UPDATE question SET formulation = ?, difficulty = ?, time = ?,\
        \ right_answer = ? WHERE question_id = ?" (formulation question,
            difficulty question:: Int, time question:: Int,
            right_answer question, getIdQuestion question)

deleteQuestion :: String -> IO ()
deleteQuestion question_id = do
    conn <- open dbPath
    execute_ conn "PRAGMA foreign_keys = ON"
    execute conn "DELETE FROM question WHERE question_id = ?" (Only question_id)

addAnswer :: String -> String -> IO ()
addAnswer textAnswer question_id = withConn dbPath $
  \conn -> do
      execute conn "INSERT INTO answer (text,question_id) \
        \VALUES (?,?)" [textAnswer,question_id]

updateAnswer:: Answer -> IO()
updateAnswer answer = do
      conn <- open dbPath
      execute conn "UPDATE answer SET text = ?  WHERE answer_id = ?"
        (text answer, getAnswerId answer)

deleteAnswer :: Int -> IO ()
deleteAnswer answer_id = do
    conn <- open dbPath
    execute_ conn "PRAGMA foreign_keys = ON"
    execute conn "DELETE FROM answer WHERE answer_id = ?" (Only answer_id)

File: projeto-prolog/controllers/questionController.pl
:- module(questionController, 
    [
        createQuestion/6,
        updateQuestionRightAnswer/2,
        getAllQuestions/2,
        updateQuestion/5,
        deleteQuestion/1,
        printQuestions/2,
        printQuestion/1
    ]).

:- use_module( library(prosqlite) ).

% dificuldade
difficulty(0,"Fácil").
difficulty(1,"Média").
difficulty(2,"Difícil").

createQuestion(Formulation, Difficulty, Duration, TypeQuestion, QuizId, QuestionId) :-
    uuid(QuestionUUID),
    format(atom(Query), "INSERT INTO question (question_id,formulation,difficulty,time,type_question,quiz_id) VALUES ('~w','~w',~d,~d,~d,'~w');",
        [QuestionUUID,Formulation,Difficulty,Duration,TypeQuestion,QuizId]),
    sqlite_query( db, Query, _),
    QuestionId = QuestionUUID.

updateQuestionRightAnswer(QuestionId, RightAnswer) :-
    format(atom(Query), "UPDATE question SET right_answer = '~w' WHERE question_id = '~w'",
        [RightAnswer, QuestionId]),
    sqlite_query( db, Query, _).

getAllQuestions(QuizId,Questions) :-
    format(atom(Query), "SELECT * from question WHERE quiz_id = '~w';",[QuizId]),
    findall( Row, sqlite_query(db, Query, Row), Questions ).

updateQuestion(QuestionId, Formulation, Difficulty, Duration, RightAnswer) :-
    format(atom(Query), "UPDATE question SET formulation = '~w', difficulty = ~d, time = ~d, right_answer = '~w' WHERE question_id = '~w';",
        [Formulation,Difficulty,Duration,RightAnswer,QuestionId]),
    sqlite_query(db, Query, _).

deleteQuestion(QuestionId) :-
    format(atom(Query), "DELETE FROM question WHERE question_id = '~w';",[QuestionId]),
    sqlite_query(db, 'PRAGMA foreign_keys = ON;',_),
    sqlite_query(db, Query, _).

printQuestion(Question) :- 
    Question = row(_,Formulation,DifficultyN,Duration,_,_,_),
    difficulty(DifficultyN,Difficulty),
    format('Enunciado: ~w, Dificuldade: ~w, Duração: ~ds\n',
        [Formulation,Difficulty,Duration]).

% formato da entrada: [row(QuestionId,Formulation,Difficulty,Duration,RightAnswer,QuizId,Type)]
printQuestions([],_).
printQuestions([H|T],Index) :-
    H = row(_,Formulation,DifficultyN,Duration,_,_,_),
    difficulty(DifficultyN,Difficulty),
    format('~d - Enunciado: ~w, Dificuldade: ~w, Duração: ~ds\n',
        [Index,Formulation,Difficulty,Duration]),
    NIndex is Index + 1, 
    printQuestions(T, NIndex).
